### PR TITLE
Unification of all test-framework-related options

### PR DESCRIPTION
**Pull request by fedinskiy**
### Summary

Since all of them are now gathered in one place
it would become easier to monitor and document them
Required for potential automation of https://github.com/quarkus-qe/quarkus-test-framework/issues/369
Future changes in test suite require accommodation of  https://github.com/quarkus-qe/quarkus-test-framework/pull/309

Please check the relevant options

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Dependency update
- [x] Refactoring
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] This change requires a documentation update
- [ ] This change requires execution against OCP (use `run tests` phrase in comment)

### Checklist:
- [ ] Example scenarios has been updated / added
- [x] Methods and classes used in PR scenarios are meaningful
- [x] Commits are well encapsulated and follow [the best practices](https://cbea.ms/git-commit/)

### CODE REVIEW

**Comment by fedinskiy** [comment]
run tests

**Comment by mocenas** [comment]
Is this PR still active / relevant @fedinskiy ?

**Comment by fedinskiy** [comment]
@mocenas this MR is relevant, but not very active.

**Comment by fedinskiy** [comment]
run tests

**Comment by fedinskiy** [comment]
run tests

**Comment by QuarkusQE** [comment]
Can one of the admins verify this patch?

**Comment by rsvoboda** [comment]
@mjurc @michalvavrik @fedinskiy what about this effort?

**Comment by michalvavrik** [comment]
> @mjurc @michalvavrik @fedinskiy what about this effort?

framework has changed a little, I think it requires additional work and I doubt @fedinskiy has time for it; let's hear from him

**Comment by fedinskiy** [comment]
I am working on this, when I have time, lately I didn't have any. Are we in a hurry with this?

**Comment by rsvoboda** [comment]
I was just wondering if this is abandoned or still on radar.

**Comment by michalvavrik** [comment]
@fedinskiy please update this PR with a property added in the https://github.com/quarkus-qe/quarkus-test-framework/pull/1467. Thank you

**Comment by fedinskiy** [comment]
run tests

**Comment by fedinskiy** [comment]
run tests

**Comment by fedinskiy** [comment]
run tests

**Comment by fedinskiy** [comment]
run tests

File: quarkus-test-core/src/main/java/io/quarkus/test/configuration/Configuration.java [hunk 1]
```diff
@@ -97,32 +111,129 @@ public static Configuration load(String... serviceNames) {
...
+    private static EnumMap<Property, String> loadPropertiesFrom(String propertiesFile, String scope) {
         try (InputStream input = Configuration.class.getClassLoader().getResourceAsStream(propertiesFile)) {
             Properties prop = new Properties();
             prop.load(input);
             return loadPropertiesFrom(prop, scope);
-        } catch (Exception ignored) {
-            // There is no properties file: this is not mandatory.
+        } catch (Exception exception) {
+            if (exception instanceof NullPointerException && exception.getMessage().equals("inStream parameter is null")) {
+                System.out.println("No properties file: " + propertiesFile);
```
Comment (michalvavrik) [inline quarkus-test-core/src/main/java/io/quarkus/test/configuration/Configuration.java:125]: I presume this is `System.out.println` because it can be called very early. That is fine. Why `System.out` and not `System.err` ?

Comment (michalvavrik) [reply quarkus-test-core/src/main/java/io/quarkus/test/configuration/Configuration.java:125]: And what is it about btw, `intStream parameter is null` ?

File: quarkus-test-core/src/main/java/io/quarkus/test/configuration/PropertyLookup.java [hunk 2]
```diff
@@ -37,7 +37,7 @@ public String get(ServiceContext service) {
         }
 
         // Or from test.properties
-        value = service.getOwner().getConfiguration().get(propertyKey);
+        value = Configuration.Property.getByName(propertyKey).map(GLOBAL::get).orElse("");
```
Comment (michalvavrik) [inline quarkus-test-core/src/main/java/io/quarkus/test/configuration/PropertyLookup.java:40]: `GLOBEL::get` is not the same thing as a configuration property from a service context, or did you change it? that could have big implications, please let me know before I'll think about it more

File: quarkus-test-core/src/main/java/io/quarkus/test/services/quarkus/LocalhostQuarkusApplicationManagedResource.java [hunk 3]
```diff
@@ -176,8 +176,10 @@ private List<String> getPropertiesForCommand() {
         if (model.isGrpcEnabled()) {
             runtimeProperties.putIfAbsent(QUARKUS_GRPC_SERVER_PORT_PROPERTY, "" + assignedGrpcPort);
         }
-
         return runtimeProperties.entrySet().stream()
+                .peek(e -> {
+                    System.out.println(e.getKey() + " " + e.getValue());
```
Comment (michalvavrik) [inline quarkus-test-core/src/main/java/io/quarkus/test/services/quarkus/LocalhostQuarkusApplicationManagedResource.java:181]: haha, let's remove that

File: quarkus-test-knative-events/root/src/main/java/io/quarkus/test/services/knative/eventing/OpenShiftExtensionFunqyKnativeEventsService.java [hunk 4]
```diff
@@ -8,9 +8,8 @@ public class OpenShiftExtensionFunqyKnativeEventsService extends FunqyKnativeEve
     @Override
     public ServiceContext register(String serviceName, ScenarioContext context) {
         // we set deployment target and registry so that every test don't have to do it (and it's in one place)
-        final String serviceScopePrefix = "ts." + serviceName;
-        System.setProperty(serviceScopePrefix + ".quarkus.kubernetes.deployment-target", "knative");
-        System.setProperty(serviceScopePrefix + ".quarkus.container-image.registry",
+        System.setProperty("quarkus.kubernetes.deployment-target", "knative");
+        System.setProperty("quarkus.container-image.registry",
```
Comment (michalvavrik) [inline quarkus-test-knative-events/root/src/main/java/io/quarkus/test/services/knative/eventing/OpenShiftExtensionFunqyKnativeEventsService.java:12]: I suspect you have to do this because of changes you made in the lookup. I am not fine with the setting properties, but then, I suppose it was me who wrote this code and it was very early, I'd probably don't do it now.

TL;DR; I can rewrite it in follow-ups, but for now, I don't think it is fine to set system properties unless we know 100 % every test class is started in new java process that is not affected by this. is it?

File: quarkus-test-service-kafka/src/main/java/io/quarkus/test/services/containers/BaseKafkaContainerManagedResource.java [hunk 5]
```diff
@@ -130,8 +131,8 @@ private void startRegistryIfEnabled() {
             if (!isRegistryRunning()) {
                 schemaRegistry.start();
             }
-
-            model.getContext().put(KafkaService.KAFKA_REGISTRY_URL_PROPERTY, getSchemaRegistryUrl());
+            System.out.println("Schema: " + getSchemaRegistryUrl());
```
Comment (michalvavrik) [inline quarkus-test-service-kafka/src/main/java/io/quarkus/test/services/containers/BaseKafkaContainerManagedResource.java:134]: well, here you can use logger I am certain that it should be ready by now, also I think maybe improve the message bit?

**Comment by michalvavrik** [comment]
I have finished first review, I'll need answers to my comments and have a second look. I think overall it looks very good and there is very little to change, but I am nervous about service-specific properties being gone, I'll wait for replies and think about it a lot.

File: quarkus-test-core/src/main/java/io/quarkus/test/configuration/Configuration.java (hunk 1 above)
Comment (fedinskiy) [reply quarkus-test-core/src/main/java/io/quarkus/test/configuration/Configuration.java]: `intStream parameter is null` means there is no properties file, this is an error thrown by one of the external classes.

Changed to `err` stream

File: quarkus-test-core/src/main/java/io/quarkus/test/services/quarkus/LocalhostQuarkusApplicationManagedResource.java (hunk 3 above)
Comment (fedinskiy) [reply quarkus-test-core/src/main/java/io/quarkus/test/services/quarkus/LocalhostQuarkusApplicationManagedResource.java]: fair point, done

File: quarkus-test-knative-events/root/src/main/java/io/quarkus/test/services/knative/eventing/OpenShiftExtensionFunqyKnativeEventsService.java (hunk 4 above)
Comment (fedinskiy) [reply quarkus-test-knative-events/root/src/main/java/io/quarkus/test/services/knative/eventing/OpenShiftExtensionFunqyKnativeEventsService.java:12]: We were setting system properties before that, just without the prefix. Could you explain, why this would be a problem now?

File: quarkus-test-core/src/main/java/io/quarkus/test/configuration/PropertyLookup.java (hunk 2 above)
Comment (fedinskiy) [reply quarkus-test-core/src/main/java/io/quarkus/test/configuration/PropertyLookup.java]: I do not remember, when did I change it, but it seems like a typo (this check is done later, in the `get()` method)

File: quarkus-test-core/src/main/java/io/quarkus/test/configuration/Configuration.java (hunk 1 above)
Comment (michalvavrik) [reply quarkus-test-core/src/main/java/io/quarkus/test/configuration/Configuration.java]: ok, thx

File: quarkus-test-knative-events/root/src/main/java/io/quarkus/test/services/knative/eventing/OpenShiftExtensionFunqyKnativeEventsService.java (hunk 4 above)
Comment (michalvavrik) [reply quarkus-test-knative-events/root/src/main/java/io/quarkus/test/services/knative/eventing/OpenShiftExtensionFunqyKnativeEventsService.java:12]: `OpenShiftExtensionFunqyKnativeEventsService` is one instance of the `Service` and each test class can contain one to many services. These properties are only valid for this service. System properties without prefix are global. Whether we have a test scenario that would currently fail or not, I want to have a separate, service-specific context specific for each Service instance. This way, we have clear boundaries between service-specific properties.

Now back to the solution, please kindly try this instead:

```
@Override
    public ServiceContext register(String serviceName, ScenarioContext context) {
        // we set deployment target and registry so that every test don't have to do it (and it's in one place)
        ServiceContext serviceContext = super.register(serviceName, context);
        serviceContext.withTestScopeConfigProperty("quarkus.kubernetes.deployment-target", "knative");
        serviceContext.withTestScopeConfigProperty("quarkus.container-image.registry",
                "image-registry.openshift-image-registry.svc:5000");
        return serviceContext;
    }
```

I have tried it with the current main and it works, maybe it will work with your implementation as well. Thanks

File: quarkus-test-core/src/main/java/io/quarkus/test/configuration/PropertyLookup.java (hunk 2 above)
Comment (michalvavrik) [reply quarkus-test-core/src/main/java/io/quarkus/test/configuration/PropertyLookup.java]: yep, thanks for the fix, it must be service specific.

**Comment by fedinskiy** [comment]
The jenkins tests are green, but they are failing, for unclear, but PR-related reasons. Will investigate further

**Comment by fedinskiy** [comment]
run tests

**Comment by michalvavrik** [comment]
frankly reminder @fedinskiy , I can see you are pushing changes and running OpenShift tests, but I still believe you must address my comment on the OpenShift Funqy regarding system properties. Thanks

**Comment by fedinskiy** [comment]
run tests

**Comment by michalvavrik** [comment]
We need @mjurc  && green OCP. Then this can be merged.

**Comment by michalvavrik** [comment]
BTW I'd also recommend to check TS, but it's no biggie, IMHO we can just fix it in follow-ups, this is big enough.

**Review by mjurc — APPROVED** [review]
LGTM

**Comment by mjurc** [comment]
Native failure unrelated; merging

File: quarkus-test-core/src/main/java/io/quarkus/test/configuration/Configuration.java [hunk 6]
```diff
@@ -97,32 +111,131 @@ public static Configuration load(String... serviceNames) {
...
+    private static EnumMap<Property, String> loadPropertiesFrom(String propertiesFile, String scope) {
         try (InputStream input = Configuration.class.getClassLoader().getResourceAsStream(propertiesFile)) {
             Properties prop = new Properties();
             prop.load(input);
             return loadPropertiesFrom(prop, scope);
-        } catch (Exception ignored) {
-            // There is no properties file: this is not mandatory.
+        } catch (Exception exception) {
+            if (exception instanceof NullPointerException && exception.getMessage().equals("inStream parameter is null")) {
+                System.err.println("No properties file: " + propertiesFile);
```
Comment (michalvavrik) [inline quarkus-test-core/src/main/java/io/quarkus/test/configuration/Configuration.java:125]: I just saw `No properties file: test.properties` when running `mvn clean verify -Dit.test=MssqlHandlerIT` in `quarkus-test-suite/sql-db/vertx-sql`. I am not on a fence, I didn't realize this can be normal situation that there are no test properties, but on the other hand, the message is not noisy and does not indicate something went wrong. We  should probably keep it as is :man_shrugging:
